src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp

```diff
@@ -206,19 +225,91 @@ APPLICATION_INFO::TryCreateHandler(
             return S_OK;
         }
     }
+
     return S_FALSE;
 }
 
 VOID
 APPLICATION_INFO::ShutDownApplication(const bool fServerInitiated)
 {
+    IAPPLICATION* app = nullptr;
+    {
+        SRWExclusiveLock lock(m_applicationLock);
+        if (!m_pApplication)
+        {
+            return;
+        }
+        app = m_pApplication.get();
+    }
+
+    LOG_INFOF(L"Stopping application '%ls'", QueryApplicationInfoKey().c_str());
+    app->Stop(fServerInitiated);
+
     SRWExclusiveLock lock(m_applicationLock);
 
-    if (m_pApplication)
+    m_pApplication = nullptr;
+    m_pApplicationFactory = nullptr;
+}
+
+std::filesystem::path
+APPLICATION_INFO::HandleShadowCopy(const ShimOptions& options, IHttpContext& pHttpContext)
+{
+    std::filesystem::path shadowCopyPath;
+
+    // Only support shadow copying for IIS.
+    if (options.QueryShadowCopyEnabled() && !m_pServer.IsCommandLineLaunch())
     {
-        LOG_INFOF(L"Stopping application '%ls'", QueryApplicationInfoKey().c_str());
-        m_pApplication->Stop(fServerInitiated);
-        m_pApplication = nullptr;
-        m_pApplicationFactory = nullptr;
+        shadowCopyPath = options.QueryShadowCopyDirectory();
+        std::wstring physicalPath = pHttpContext.GetApplication()->GetApplicationPhysicalPath();
+
+        // Make shadow copy path absolute.
+        if (!shadowCopyPath.is_absolute())
+        {
+            shadowCopyPath = std::filesystem::absolute(std::filesystem::path(physicalPath) / shadowCopyPath);
+        }
+
+        // The shadow copy directory itself isn't copied to directly.
+        // Instead subdirectories with numerically increasing names are created.
+        // This is because on shutdown, the app itself will still have all dlls loaded,
+        // meaning we can't copy to the same subdirectory. Therefore, on shutdown,
+        // we create a directory that is one larger than the previous largest directory number.
```

Comment:
Does the order really matter? Why not use a GUID instead?

How are the directories cleaned up?

Reply:
Yeah, between process restarts, two folders will exist. We need a way to know which folder we should use between them. So a higher version is the one we should use.

Directories that aren't in use in the shadow copy base will be cleaned up on each process start on a background thread.